src/core/FObject.js

```diff
@@ -169,36 +185,70 @@ foam.CLASS({
-        // FUTURE: add benchmark to justify
-        // ???: optional exception trapping, benchmark
-        switch ( a.length ) {
-          case 0: l(s); break;
-          case 1: l(s, a[0]); break;
-          case 2: l(s, a[0], a[1]); break;
-          case 3: l(s, a[0], a[1], a[2]); break;
-          case 4: l(s, a[0], a[1], a[2], a[3]); break;
-          case 5: l(s, a[0], a[1], a[2], a[3], a[4]); break;
-          case 6: l(s, a[0], a[1], a[2], a[3], a[4], a[5]); break;
-          case 7: l(s, a[0], a[1], a[2], a[3], a[4], a[5], a[6]); break;
-          case 8: l(s, a[0], a[1], a[2], a[3], a[4], a[5], a[6], a[7]); break;
-          case 9: l(s, a[0], a[1], a[2], a[3], a[4], a[5], a[6], a[7], a[8]); break;
-          default: l.apply(l, [s].concat(Array.from(a)));
-        }
+        this.notifyListener_(l, s, a);
         listeners = listeners.next;
         count++;
       }
       return count;
     },
 
+    function notifyListener_(l, s, a) {
+      // Like l.apply(l, [s].concat(Array.from(a))), but faster.
+      // FUTURE: add benchmark to justify
+      // ???: optional exception trapping, benchmark
+      switch ( a.length ) {
+      case 0: l(s); break;
+      case 1: l(s, a[0]); break;
+      case 2: l(s, a[0], a[1]); break;
+      case 3: l(s, a[0], a[1], a[2]); break;
+      case 4: l(s, a[0], a[1], a[2], a[3]); break;
+      case 5: l(s, a[0], a[1], a[2], a[3], a[4]); break;
+      case 6: l(s, a[0], a[1], a[2], a[3], a[4], a[5]); break;
+      case 7: l(s, a[0], a[1], a[2], a[3], a[4], a[5], a[6]); break;
+      case 8: l(s, a[0], a[1], a[2], a[3], a[4], a[5], a[6], a[7]); break;
+      case 9: l(s, a[0], a[1], a[2], a[3], a[4], a[5], a[6], a[7], a[8]); break;
+      default: l.apply(l, [s].concat(Array.from(a)));
+      }
+    },
+
     function hasListeners(/* args */) {
       /** Return true iff there are listeners for the supplied message. **/
+      var args = Array.from(arguments);
+      return this.hasKeyedListeners_(args) || this.hasExprListeners_(args);
+    },
+
+    function hasKeyedListeners_(args) {
+      /** Return true iff there are string-keyed listeners for message. **/
       var listeners = this.getPrivate_('listeners');
 
-      for ( var i = 0 ; listeners ; i++ ) {
-        if ( listeners.next        ) return true;
-        if ( i == arguments.length ) return false;
-        listeners = listeners.children && listeners.children[arguments[i]];
+      for ( var i = 0; listeners; i++ ) {
+        if ( listeners.next   ) return true;
+        if ( i == args.length ) return false;
+        if ( typeof args[i] === 'string' ) {
+          listeners = listeners.children && listeners.children[args[i]];
+        }
+      }
+
+      return false;
+    },
+
+    function hasExprListeners_(args) {
+      /** Return true iff there are expr-matched listeners for message. **/
+      var listeners = this.getPrivate_('listeners');
+
+      for ( var i = 0; listeners; i++ ) {
+        if ( listeners.hasExprs ) {
```

Comment:
@kgrgreer, that is my intention, yes. It serves the only use case we have so far. To clarify, don't most listeners include matchable data in the topic? E.g., most common event topic looks like ['propertyChange', someName, someObj], so I can use .sub('propertyChange', 'foo', myExpr) for a filtered subscription to foo-property-change.s